Inject into `KItem` instead of `K`

nwatson22 opened the pull request:
closes #565 
We weren't doing the kast to kore conversion in the same way as `kore-exec`, resulting in a possibly invalid kore term (per discussion on Slack involving @goodlyrottenapple @tothtamas28 and @virgil-serbanuta).
This changes the conversion so that the injection is not done directly into the `K` sort, but into a `KItem` which is embedded into a K sequence.
I don't know much about kore so I'm not sure if this is correct, but it did produce the correct expected output on #565.

tothtamas28 commented on src/pyk/kore/kompiled.py:
```diff
@@ -75,10 +75,10 @@ def is_subsort(self, sort1: Sort, sort2: Sort) -> bool:
 
-        if sort2 == SortApp('SortK'):
+        if sort2 == SortApp('SortKItem'):
             return True
 
-        if sort1 == SortApp('SortK'):
+        if sort1 == SortApp('SortKItem'):
             return False
```
`SortK` is the top sort, this shouldn't be changed.

tothtamas28 replied:
At least I think it is the top sort. @radumereuta, can you confirm?

ehildenb replied:
I somehow agree this sholudn't be changed either. Fro example, if `sort1` is `SortK`, then this code will return that `SortKItem` is a subsort of `SortK`.

nwatson22 replied:
This is reverted. It now adds the KSequences while it is still in Kast form.

tothtamas28 commented on src/pyk/kore/kompiled.py:
```diff
@@ -122,4 +122,7 @@ def _inject(self, pattern: Pattern, sort: Sort) -> Pattern:
         if self.is_subsort(actual_sort, sort):
             return App('inj', (actual_sort, sort), (pattern,))
 
+        if sort == SortApp('SortK'):
```
Please add a comment for this special case.

Also this branch needs to precede the subsort checking.

nwatson22 replied:
This case has been removed and now the KSeqences are added earlier, when the term is still in Kast format. Added a comment for this conversion step.

tothtamas28 commented:
I wonder though, is there a reason why `SortK` doesn't admit injections? Maybe the ultimate solution would be to just generated the corresponding axioms on `inj` based on

```
syntax K ::= KItem
```

@radumereuta @virgil-serbanuta

virgil-serbanuta commented:
> I wonder though, is there a reason why `SortK` doesn't admit sort injections? 

I think that `K` is supposed to be a list of `KItem`, based on two constructors, `kseq` and `dotk`. This is used as such most obviously in the `<k>` cell.

If you would allow injections, that would make things ambiguous, or would make unification hard, or both.

The main question is in what way an injection is related to a single element list.

If an injection should be treated as a single element list, that would mean that `inj{KItem, K}(stuff)` should unify with `kseq(stuff, dotk)`, which would mean that `kseq` is no longer a constructor, which would make unification harder than it should be.

If an injection is not the same as a single element list, then a rule containing `<k>stuff</k>`, i.e., containing `<k>stuff ~> .K </k>`, would not unify with a configuration containing `<k>inj(stuff)</k>`, which would probably be confusing.

virgil-serbanuta commented on src/pyk/kore/kompiled.py:
```diff
@@ -122,4 +122,7 @@ def _inject(self, pattern: Pattern, sort: Sort) -> Pattern:
             return App('inj', (actual_sort, sort), (pattern,))
 
+        if sort == SortApp('SortK'):
+            return App('kseq', (), (App('inj', (actual_sort, SortApp('SortKItem')), (pattern,)), App('dotk')))
```
Would it be reasonable to use `KSequence` and `_ksequence_to_kore` here?

ehildenb replied:
This makes sense. Basically in `kast_to_kore`, when we see something of sort `KItem` being used in a place where sort `K` is expected, emit `_ksequence_to_kore(KSequence[KITEM])`.

Maybe then we don't need to touch anything here.

nwatson22 replied:
Moved the whole conversion to before the `_kast_to_kore` conversion, so we are now using `KSequence`.

virgil-serbanuta commented:
> @virgil-serbanuta perhaps a more sensible change would be to have `==K` defined between two `KItems`? Tho this would be a very pervasive change too, so not sure what the consequences would be

I think that `==K` is currently used for actual sequences of `KItem`, so it should stay that way. Also, `==Sort`, whenever it's defined, means equality on `Sort`, so changing that for `K` would be confusing.

ehildenb commented:
Let's add unit tests of this (or integration tests) of `kast_to_kore`.

But in general, `KItem` is not the top sort, `K` is. So we shouldn't treat `KItem` as such. we should be special casing the handful of places where we need to do this injection.

What does the original term we have look like? And what the backend expects?

goodlyrottenapple commented:
@ehildenb from the linked issue:

pyk does not encode the `==K` predicates correctly into kore. Currently, for `LHS ==K RHS`, we emit:

```
Lbl'UndsEqlsEqls'K'Unds'{}(
    inj{SortInt{}, SortK{}}(LHS),
    inj{SortInt{}, SortK{}}(RHS)
)
```

however, the correct encoding should be

```
Lbl'UndsEqlsEqls'K'Unds'{}(
  kseq{}(
    inj{SortInt{}, SortKItem{}}(ConfigVarX:SortInt{}),
    dotk{}()
  ),
  kseq{}(
    inj{SortInt{}, SortKItem{}}(\dv{SortInt{}}("0")),
    dotk{}()
  )
)
```

nwatson22 commented:
> Ok, then I think we should probably special case `_==K_` in `kast_to_kore`, to insert the `KSequence` underneath it. It can be added as a pass here:
> 
> https://github.com/runtimeverification/pyk/blob/1679ce41cd9b8c281a2ddda5295e3d37a8a54f48/src/pyk/konvert.py#L90
> 
> Basically, before that line, traverse the configuration and do `add_ksequence_under_kequal` to the kast term.
> 
> But, more importantly, add a test of this directly first, for example here:
> 
> https://github.com/runtimeverification/pyk/blob/1679ce41cd9b8c281a2ddda5295e3d37a8a54f48/src/tests/integration/konvert/test_simple_proofs.py#L473

Added this as a test and changed to be a special case of `_==K_`.

ehildenb commented:
This looks good to me. Would be good to confirm that it solves the original issue as well before adding automerge.

virgil-serbanuta commented on src/pyk/kast/outer.py:
```diff
@@ -1137,6 +1137,16 @@ def greatest_common_subsort(self, sort1: KSort, sort2: KSort) -> KSort | None:
+    def add_ksequence_under_kequal(self, kast: KInner) -> KInner:
+        def _add_ksequence_under_kequal(kast: KInner) -> KInner:
+            if type(kast) is KApply and kast.label.name == '_==K_':
+                return KApply('_==K_', [KSequence(arg) for arg in kast.args])
```
What happens if `arg` was already a `KSequence`?

ehildenb replied:
Yes, `KSequence` is able to flatten, but a test wolud be good.

ehildenb replied:
I guess it would be good to also test for variables of sort `K` underneath the equals, to make sure they _don't_ get injected into the KSequence.

And to test for empty KSequence and non-empty KSequence under the equals.

nwatson22 commented:
I confirmed this does fix the issue that originally appeared in KEVM.
